File: src/gordon_hardware/src/rotary_encoder.cpp
#include <pigpiod_if2.h>
#include <stdint.h>

#include <atomic>
#include <stdexcept>

#include "gordon_hardware/rotary_encoder.hpp"

RotaryEncoder::RotaryEncoder(uint8_t a_pin, uint8_t b_pin)
    : count(0)
    , a_pin(a_pin)
    , a_value(0)
    , b_pin(b_pin)
    , b_value(0)
    , previous_pin(0)
{
    pi = pigpio_start(nullptr, nullptr);
    if (pi < 0) {
        throw std::runtime_error("Failed to initialize pigpio");
    }

    set_mode(pi, a_pin, PI_INPUT);
    set_mode(pi, b_pin, PI_INPUT);

    set_pull_up_down(pi, a_pin, PI_PUD_UP);
    set_pull_up_down(pi, b_pin, PI_PUD_UP);

    a_callback = callback_ex(pi, a_pin, EITHER_EDGE, callbackEx, this);
    b_callback = callback_ex(pi, b_pin, EITHER_EDGE, callbackEx, this);
}

RotaryEncoder::~RotaryEncoder()
{
    callback_cancel(a_callback);
    callback_cancel(b_callback);

    pigpio_stop(pi);
}

void RotaryEncoder::callbackEx(int pi, unsigned pin, unsigned value, uint32_t tick, void* data)
{
    RotaryEncoder* rotary_encoder = reinterpret_cast<RotaryEncoder*>(data);
    rotary_encoder->callback(pi, pin, value, tick);
}

void RotaryEncoder::callback([[gnu::unused]] int pi, unsigned pin, unsigned value, [[gnu::unused]] uint32_t tick)
{
    if (pin == a_pin) {
        a_value = value;
    } else if (pin == b_pin) {
        b_value = value;
    }

    if (pin != previous_pin) {
        previous_pin = pin;

        if (pin == a_pin && value == 1) {
            if (b_value == 1) {
                count++;
            }
        } else if (pin == b_pin && value == 1) {
            if (a_value == 1) {
                count--;
            }
        }
    }
}

File: src/gordon_control/gordon_control/drivers/rotary_encoder.py
import pigpio


class RotaryEncoder:
    @property
    def count(self) -> int:
        return self._count

    _pi: pigpio.pi

    _a_pin: int
    _b_pin: int

    _previous_pin: int = 0

    _a_value: int = 0
    _b_value: int = 0

    _count: int = 0

    def __init__(self, a_pin: int, b_pin: int) -> None:
        self._a_pin = a_pin
        self._b_pin = b_pin

        self._pi = pigpio.pi()
        if not self._pi.connected:
            raise RuntimeError("Failed to initialize GPIO")

        self._pi.set_mode(self._a_pin, pigpio.INPUT)
        self._pi.set_mode(self._b_pin, pigpio.INPUT)

        self._pi.set_pull_up_down(self._a_pin, pigpio.PUD_DOWN)
        self._pi.set_pull_up_down(self._b_pin, pigpio.PUD_DOWN)

        self._pi.callback(self._a_pin, pigpio.EITHER_EDGE, self._callback)
        self._pi.callback(self._b_pin, pigpio.EITHER_EDGE, self._callback)

    def close(self) -> None:
        self._pi.set_mode(self._a_pin, pigpio.OUTPUT)
        self._pi.set_mode(self._b_pin, pigpio.OUTPUT)

        self._pi.write(self._a_pin, 0)
        self._pi.write(self._b_pin, 0)

    def _callback(self, pin: int, value: int, _: int) -> None:
        if pin == self._a_pin:
            self._a_value = value

        elif pin == self._b_pin:
            self._b_value = value

        if not pin == self._previous_pin:
            self._previous_pin = pin

            if pin == self._a_pin and value == 1:
                if self._b_value == 1:
                    self._count -= 1

            elif pin == self._b_pin and value == 1:
                if self._a_value == 1:
                    self._count += 1
